lib/chef/resource_inspector.rb

```diff
@@ -69,6 +73,20 @@ def self.extract_resource(resource, complete = false)
       data
     end
 
+    def self.resolve_action_description(resource, action)
+      resource.action_description(action) ||
+        resource.new(resource.name, nil).provider_for_action(action).class.action_description(action)
+    rescue NameError => e
+      # Some providers will only properly load when the platform is correct -
+      # so for exmaple, we'll fail on some Windows resources because we don't
+      # invoke Windows platform-specific requires when not on Windows.
+      raise unless e.message =~ /uninitialized constant/
+    rescue Chef::Exceptions::ProviderNotFound
+      # Providers that are platform-specific won't correctly resolve,
+      # because we don't provide any platform information to resolve them.
+      nil
+    end
```

Comment:
you should probably push this down into the resource, and keep the `resource.action_description(action)` call here in the ResourceInspector.

it is also a bit weird to have the action_description on both the resource itself and on the provider.  the resource should have a facade for action_description which should then always construct and call the underlying provider for the real data.

Reply:
```
    def self.action(action, description: nil, &recipe_block)
      action = action.to_sym
      declare_action_class
      action_class.action(action, &recipe_block)
```

in the resource that last line should just be `action_class.action(action, description &recipe_block)` to delegate management of the action_description to the provider/action_class.

when retrieving the action_description you are no longer at class parsing time, and the resource has been built and the system knows how to find the right provider based on if it wes declared as a custom resource or LWRP, so you can delegate to the provider through `provider_for_action` and the magic is taken care of.

if you added some other declaration e.g.:

```
action_description :foo, "whatever"
action :foo do
  ...
end
```

then action_description would only be valid in a resource file if it was a custom resource and in an LWRP would only be valid on the provider.  so then the action_description definition in the resource should do `declare_action_class` and then delegate to the `action_class.action_description`.  (gets a bit trickier with a method that as a writer would only be valid against a custom resource, but as a reader could be used against an LWRP, but just being careful about the conditional to determine how it was called would make that still doable)